### ISSUE TITLE
Render widgets based on AppLifeCycleState is not working (build method is never called for inactive state)

### ISSUE THREAD

**Issue by pravinarr** [flutter/flutter#33236]
I have the following code which will replace the child with a container when the app goes to inactive or paused state.

```

import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        body: SecureScreen(
          child: Container(
            color: Colors.red,
          ),
        ),
      ),
    );
  }
}

class SecureScreen extends StatefulWidget {
  final Widget child;
  SecureScreen({@required this.child});

  @override
  SecureScreenState createState() => SecureScreenState();
}

class SecureScreenState extends State<SecureScreen>
    with WidgetsBindingObserver {
  AppLifecycleState _notification;

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addObserver(this);
  }

  @override
  void dispose() {
    WidgetsBinding.instance.removeObserver(this);
    super.dispose();
  }

  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    print(state.toString());
    setState(() {
      _notification = state;
    });
  }

  @override
  Widget build(BuildContext context) {
    return (_notification == AppLifecycleState.paused ||
            _notification == AppLifecycleState.inactive)
        ? Container(
            color: Colors.greenAccent,
          )
        : widget.child;
  }
}


```

So when I double click the home button in IOS to view all the open applications, it is triggering
"inactive" state but it is not calling the build() method to render container with greenAccent color.

I have attached the screenshot for you reference.

Does this needs additional code to achieve the result ? 
![image](https://user-images.githubusercontent.com/8682635/58225618-b0f26800-7cd7-11e9-958d-6f9d96ddd257.png)

**Comment by pravinarr** [comment]
Related #27344

**Comment by dnfield** [comment]
I suspect what we want here is something in `viewWillDisappear` getting pumped down to the framework somehow.  I don't think we can reliably expect anything to execute in a paused/inactive state for drawing on the screen.

**Comment by dnfield** [comment]
Per input from @matthew-carroll marking this as critical.  Customer(s) who want to hide what's on the screen before going to background need this.

**Comment by matthew-carroll** [comment]
Triage note: This issue prevents security conscious apps from clearing sensitive UI from apps that are moved to background.

**Comment by dnfield** [comment]
I'm not sure that there's goin got be a great way to do this in Dart - particularly in a way that would guarantee that a message arrived in Dart land in time for Flutter to render a new frame with the desired content.

However, you can do this pretty trivially in a Flutter app.  On iOS, following https://developer.apple.com/library/archive/qa/qa1838/_index.html, I changed up `AppDelegate.m` in a sample project to be like this:

```objc
#include "AppDelegate.h"
#include "GeneratedPluginRegistrant.h"

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application
    didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
  [GeneratedPluginRegistrant registerWithRegistry:self];
  // Override point for customization after application launch.
  return [super application:application didFinishLaunchingWithOptions:launchOptions];
}

- (void)applicationDidEnterBackground:(UIApplication *)application
{
    // Your application can present a full screen modal view controller to
    // cover its contents when it moves into the background. If your
    // application requires a password unlock when it retuns to the
    // foreground, present your lock screen or authentication view controller here.

    UIViewController *blankViewController = [UIViewController new];
    blankViewController.view.backgroundColor = [UIColor blackColor];

    // Pass NO for the animated parameter. Any animation will not complete
    // before the snapshot is taken.
    [self.window.rootViewController presentViewController:blankViewController animated:NO completion:NULL];
}

- (void)applicationWillEnterForeground:(UIApplication *)application
{
    // This should be omitted if your application presented a lock screen
    // in -applicationDidEnterBackground:
    [self.window.rootViewController dismissViewControllerAnimated:NO completion:NO];
}

@end
```

And now I get a black screen in the switcher.

**Comment by pravinarr** [comment]
@dnfield Thank you for providing workaround for this issue. It worked perfectly. Can you provide some guidance to achieve the same on android?

**Comment by dnfield** [comment]
You can add  `getWindow().addFlags(WindowManager.LayoutParams.FLAG_SECURE);` to `onCreate` of `MainActivity.java`, which will also end up preventing screen shots from working. If you want it only to work when the app gets backgrounded, you could instead do something like 

```java
    @Override
    public void onResume() {
        super.onResume();
        getWindow().clearFlags(WindowManager.LayoutParams.FLAG_SECURE);
    }

    @Override
    public void onPause() {
        super.onPause();
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_SECURE);
    }
```

Although I'm not 100% sure this would end up covering something like an embedded platform view (e.g. a webview in your flutter app).

**Comment by dnfield** [comment]
Based on the work arounds available for this, I'm going to remove the critical label. Please let me know (or feel free to re-add it if you're able) if this needs more attention again.

In case it's not already clear, I think it's unlikely that we'll be able to do a whole lot here on the framework side for this, but perhaps there's some good way to turn this into a plugin of some sort.

**Comment by matthew-carroll** [comment]
@dnfield I think it would probably be a good idea to create a plugin for this. I'll follow up with you and others offline to see what makes sense.

**Comment by Kavantix** [comment]
@dnfield Native iOS apps actually continue to render when inactive.
Only when going to the background they stop rendering.
This is especially noticeable when there is a permission alert, a native app would keep rendering so for instance an activity indicator keeps animating. In flutter however a circularprogressindicator will stop.
Also when going into multitasking, native apps keep updating their ui, flutter apps stop immediately.

**Comment by LinusU** [comment]
Same thing that I reported in #32895

**Comment by danspano** [comment]
I agree with @Kavantix on this point. It is unexpected behavior for an iOS app to stop rendering in this state.

Also, Flutter 1.0 actually did render frames while inactive (you can test this on the 1.0 branch). Only after upgrading from Flutter 1.0 to the next stable release, I noticed this behavior changed and the app stopped rendering in inactive state. I tried to find information in the release notes but couldn't find anything that suggests this was an intentional change, but seems major enough to warrant a mention.

Hoping to see this addressed in a future release as it impacts quite a few use cases. There are many reasons why a system dialog may appear (e.g request FaceID/TouchID), when these appear, any animation, navigation transition, etc. is frozen and it looks very odd.

**Comment by mengzhidaren** [comment]
Has this problem been solved? I have also encountered this problem.

**Comment by hariom08** [comment]
> I suspect what we want here is something in `viewWillDisappear` getting pumped down to the framework somehow. I don't think we can reliably expect anything to execute in a paused/inactive state for drawing on the screen.

@matthew-carroll , @dnfield - As you say, we can't reliably expect anything to execute in the paused state. But for me,  because of this issue, I'm in a deadlock and I can't submit the app to the app store. In my case, the UI and data related functionalities are split with BLOC pattern and handled through streams and streambuilder. Although my UI related workflow will not get affected much because of this issue, my data is seriously affected. I don't have a way to trigger methods that resides in another class which can be accessed only via stream from the UI, for which I need buildcontext and build method to be called. 
I wonder if there is any workaround to do this by not changing the existing BLOC pattern drastically. By the way, the app works fine in android. 

Also, let me know if I should create this as a new issue because I think this is critical.

**Comment by matthew-carroll** [comment]
@hariom08 I'm a bit confused about the issue you're describing. You mention BLoCs that need to execute behavior, but I'm confused as to why they can't continue to execute regardless of Flutter's state? 

You mentioned that there are streams from the widget tree to your BLoCs, but the widget tree is a UI construct. It represents UI interactions. So there shouldn't be any events expected from the widget tree when Flutter is paused, right?

**Comment by hariom08** [comment]
@matthew-carroll - My situation is this while developing the idea to share the app data to the UI, I preferred creating my data part in a separate class and use streams, BLoC pattern and inherited widget to access the data to and from the UI.  This being the case, to access the data part from the UI, I need the buildcontext from the build method to access the inheritedwidget. 

>  So there shouldn't be any events expected from the widget tree when Flutter is paused, right?

Yes, you are right. No UI event is expected from the UI when the app is paused. But in my case, to know the app's paused state, I make use of didChangeAppLifecycleState of WidgetsBindingObserver in a stateful widget and use this to trigger the build method with the setState call. The build is called if and only the app's state is paused. Though this setState method is not going to update the UI, I use this as a hack to trigger and execute methods in the data class.

The point is, using bloc pattern via inheritedwidget to access data, or using didChangeAppLifecycleState to know the app's state doesn't seem to be inappropriate. But now for my situation, I'm stuck as to how else would I access the methods having implemented the idea that works fine in android but not in iOS. 
I'm open to change and also looking for a meaningful workaround if any. And thanks for your quick response.

**Comment by matthew-carroll** [comment]
@hariom08 would it make sense to simply cache a reference to your BLoC within that StatefulWidget? I think within `didChangeDependencies()` you can probably access the `InheritedWidget` and obtain a reference to your BLoC. Then you can store it in a local variable in your `StatefulWidget`. When your `WidgetsBinding` callback gets invoked for `didChangeAppLifecycleState` you could directly invoke any method you'd like on your BLoC. Would that solve the problem you're facing?

**Comment by hariom08** [comment]
@matthew-carroll thank you for your suggestion. It really helped to solve my current problem. But what I was looking for is some sort of an interface that deals with applifecyclestate without binding it with the widget. Not sure what I'm asking is a meaningful request considering what flutter is made for.

**Comment by ajviera** [comment]
Any solution to this error? I don't think the feature label is correct, this is a bug in the handling of the AppLifecycleState.

**Comment by matthew-carroll** [comment]
@hariom08 would implementing a `WidgetsBindingObserver` solve your goal of reacting to lifecycle changes outside the widget tree?
https://api.flutter.dev/flutter/widgets/WidgetsBindingObserver/didChangeAppLifecycleState.html

**Comment by matthew-carroll** [comment]
@jmagman do you happen to have any iOS lifecycle insights as per the issue in this ticket? Based on the thread, it seems like maybe Flutter's rendering process needs to be extended with regard to iOS lifecycle phases, or that a new phase concept needs to be added to Flutter. Do you happen to have any thoughts on this?

**Comment by jmagman** [comment]
As you guys already explored, it would be easy to do this in a plugin since the `FlutterApplicationLifeCycleDelegate ` already exposes [`-applicationDidEnterBackground` and `-applicationWillEnterForeground`](https://github.com/flutter/engine/blob/9c00c26c8fe85915d9cbeb88560678fc44fe7752/shell/platform/darwin/ios/framework/Headers/FlutterPlugin.h#L59).

As soon as [`-[FlutterViewController applicationDidEnterBackground :]`](https://github.com/flutter/engine/blob/23df3e6bcbcbbd4b475249f94978e66c6be127f5/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm#L552) returns it will let the app actually be backgrounded.  If we wanted to allow some dart code to run, that method would have to wait until the dart code has run, THEN return.  And [from the docs](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1622997-applicationdidenterbackground), that code will have ~5 seconds:

> Your implementation of this method has approximately five seconds to perform any tasks and return. ... In practice, you should return from applicationDidEnterBackground(_:) as quickly as possible. If the method does not return before time runs out your app is terminated and purged from memory.

So `[[_engine.get() lifecycleChannel] sendMessage:state];` would need to handle some callback on the channel to say that the lifecycle message has been handled (I'm not familiar with that code).

**Comment by gabdsg** [comment]
Is there any update on this?

**Comment by hariom08** [comment]
@matthew-carroll 



> would implementing a `WidgetsBindingObserver` solve your goal of reacting to lifecycle changes outside the widget tree?

Please check the conversion we had earlier. I initially started with WidgetsBindingObserver and its callback didChangeAppLifecycleState to catch the lifecycle changes, but this couldn't trigger codes outside the widget tree without calling the build method with setState call. But this hack to call codes outside the widget tree worked well in android but failed in iOS. Though didChangeAppLifecycleState() will be called in iOS, it fails to trigger the build method and the app enters background abruptly. So for this inconsistency, I was not able to execute codes outside the widget tree from within the build method, only on the iOS side.

But your suggestion to use didChangeDependencies() helped to cache a copy of inheritedwidget locally in a statefulwidget and run code outside the widget tree.

I did reply to you immediately on 7 dec 2019 :

> thank you for your suggestion. It really helped to solve my current problem.

**Comment by audente** [comment]
I had a problem similar to @pravinarr: [OP](https://github.com/flutter/flutter/issues/33236#issue-447445060).

1. I solved the Android part using the code from @dnfield. I just needed to migrate to Kotlin.
2. For the iOS version I changed the events from [this code](https://github.com/flutter/flutter/issues/33236#issuecomment-503783281), in order to use WillResignActive and DidBecomeActive. My final code is below and now both versions work as I needed them.

```
@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate {
  var blankViewController : UIViewController? = nil

  override func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
  ) -> Bool {
    GeneratedPluginRegistrant.register(with: self)
    return super.application(application, didFinishLaunchingWithOptions: launchOptions)
  }

  override func applicationWillResignActive(_ application: UIApplication){
    blankViewController = UIViewController()
    blankViewController!.view.backgroundColor = UIColor.white

    blankViewController!.modalPresentationStyle = .fullScreen
    self.window?.rootViewController?.present(blankViewController!, animated:false, completion:nil)
  }

  override func applicationDidBecomeActive(_ application: UIApplication) {
    if blankViewController != nil {  
      self.window?.rootViewController?.dismiss(animated:false, completion:nil)
      blankViewController = nil
    }
  }
}
```

**Comment by kasvith** [comment]
Any updates?

**Comment by lzhuor** [comment]
@audente Tried your solution and it works. Thanks!

**Comment by gabdsg** [comment]
Any updates?

**Comment by michael-ottink** [comment]
Any updates?

**Comment by muarachmann** [comment]
The [Flutter Application](https://api.flutter.dev/javadoc/io/flutter/app/FlutterApplication.html) doesn't override methods like `onResume`, `onPause`, is there anyway to by pass this? Also tried getting  `getCurrentActivity().onPause(....)` but not working says its protected. I will appreciate any help thanks

**Comment by muarachmann** [comment]
> The [Flutter Application](https://api.flutter.dev/javadoc/io/flutter/app/FlutterApplication.html) doesn't override methods like `onResume`, `onPause`, is there anyway to by pass this? Also tried getting `getCurrentActivity().onPause(....)` but not working says its protected. I will appreciate any help thanks

Used `Application.ActivityLifecycleCallbacks` and `registerActivityLifecycleCallbacks(listener)` then I could access the `onActivityResumed()` and `onActivityPaused()`

**Comment by Jethro87** [comment]
@audente Would you mind posting your Kotlin code? Your swift code was super helpful, thanks 👍

**Comment by Jethro87** [comment]
Managed to get it working. Reference for those that come across this issue:

```
import androidx.annotation.NonNull;
import io.flutter.embedding.android.FlutterFragmentActivity;
import io.flutter.embedding.engine.FlutterEngine;
import io.flutter.plugins.GeneratedPluginRegistrant;
import android.view.WindowManager;
import android.os.Bundle;

class MainActivity: FlutterFragmentActivity() {
    override fun configureFlutterEngine(@NonNull flutterEngine: FlutterEngine) {
        GeneratedPluginRegistrant.registerWith(flutterEngine);
    }
    override fun onCreate(savedInstanceState : Bundle?) {
        super.onCreate(savedInstanceState);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_SECURE);    
    }
}
```

Note that this is using `FlutterFragmentActivity` instead of `FlutterActivity`. Cheers.

**Comment by adeolaex** [comment]
> Managed to get it working. Reference for those that come across this issue:
> 
> ```
> import androidx.annotation.NonNull;
> import io.flutter.embedding.android.FlutterFragmentActivity;
> import io.flutter.embedding.engine.FlutterEngine;
> import io.flutter.plugins.GeneratedPluginRegistrant;
> import android.view.WindowManager;
> import android.os.Bundle;
> 
> class MainActivity: FlutterFragmentActivity() {
>     override fun configureFlutterEngine(@NonNull flutterEngine: FlutterEngine) {
>         GeneratedPluginRegistrant.registerWith(flutterEngine);
>     }
>     override fun onCreate(savedInstanceState : Bundle?) {
>         super.onCreate(savedInstanceState);
>         getWindow().addFlags(WindowManager.LayoutParams.FLAG_SECURE);    
>     }
> }
> ```
> 
> Note that this is using `FlutterFragmentActivity` instead of `FlutterActivity`. Cheers.

Thanks for this. I was able to get it working.

 Just to clarify, on IOS the result of the swift code from @audente and @dnfield , will only be able to show us a blank screen when switching apps ?

**Comment by gaaclarke** [comment]
Fixed in https://github.com/flutter/engine/pull/25196

**Comment by fribeau** [comment]
It seems that the same issue is also happening on Android.
I just copied-pasted the same exact code shown at the beginning of this issue (minus some question marks in order to work with null safety) and I have the same unwanted behavior: the background stays always `red` no matter the state.

Does the fixed made 12 days ago works only for iOS or should I try another approche than `WidgetsBindingObserver` to reach the same goal ?

This is the output of my `flutter doctor` (I work with a Pixel 5):
```bash
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, 2.1.0-13.0.pre.574, on Linux, locale fr_FR.UTF-8)
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
[✓] Chrome - develop for the web
[✓] Linux toolchain - develop for Linux desktop
[✓] Android Studio (version 4.1)
[✓] Android Studio
[✓] IntelliJ IDEA Community Edition (version 2019.3)
[✓] VS Code (version 1.55.0)
[✓] Connected device (3 available)

• No issues found!
```

**Comment by gaaclarke** [comment]
@fribeau can you file another issue for android, please?  If the steps are the same, just copy and paste them and link this issue.

The fix will only work for iOS, there isn't a cross platform solution since it's tightly tied to the embedder.
